src/Agent.Plugins/TestFilePublisher/Plugin/TestFilePublisherLogPlugin.cs

```diff
@@ -0,0 +1,207 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Agent.Plugins.Log.TestResultParser.Contracts;
+using Agent.Sdk;
+using Pipelines = Microsoft.TeamFoundation.DistributedTask.Pipelines;
+
+namespace Agent.Plugins.Log.TestFilePublisher
+{
+    public class TestFilePublisherLogPlugin : IAgentLogPlugin
+    {
+        /// <inheritdoc />
+        public string FriendlyName => "TestFilePublisher";
+
+        public TestFilePublisherLogPlugin()
+        {
+            // Default constructor
+        }
+
+        /// <summary>
+        /// For UTs only
+        /// </summary>
+        public TestFilePublisherLogPlugin(ITraceLogger logger, ITelemetryDataCollector telemetry, ITestFilePublisher testFilePublisher)
+        {
+            _logger = logger;
+            _telemetry = telemetry;
+            _testFilePublisher = testFilePublisher;
+        }
+
+        /// <inheritdoc />
+        public async Task<bool> InitializeAsync(IAgentLogPluginContext context)
+        {
+            try
+            {
+                _logger = _logger ?? new TraceLogger(context);
+                _telemetry = _telemetry ?? new TelemetryDataCollector(new ClientFactory(context.VssConnection), _logger);
+
+                await PopulatePipelineConfig(context);
+
+                _telemetry.AddOrUpdate(TelemetryConstants.PluginInitialized, true);
+                _telemetry.AddOrUpdate(TelemetryConstants.PluginDisabled, true);
+
+                if (DisablePlugin(context))
+                {
+                    return false; // disable the plugin
+                }
+
+                _testFilePublisher = _testFilePublisher ??
+                                     new TestFilePublisher(context.VssConnection, PipelineConfig, new TestFileTraceListener(context), _logger, _telemetry);
+                await _testFilePublisher.InitializeAsync();
+            }
+            catch (Exception ex)
+            {
+                context.Trace(ex.ToString());
+                _logger?.Warning($"Unable to initialize {FriendlyName}.");
+                _telemetry?.AddOrUpdate(TelemetryConstants.InitializeFailed, ex);
+                return false;
+            }
+            finally
+            {
+                if (_telemetry != null)
+                {
+                    await _telemetry.PublishCumulativeTelemetryAsync();
+                }
+            }
```

Comment:
this should not be here. It should be when we disable plugin and in a finally block in the finalize method